### PR TITLE
[RFC] Add Xen support 

### PR DESCRIPTION
**Pull request by lorc**
Hello.

This PR adds possibility to run zephyr as aarch64 Xen guest.

It includes a number of generic patches, that fix different issues and couple of Xen-specific changes.

We use Zephyr internally to debug RT-related things in Xen and maybe community will be interested in running Zephyr as Xen guest.

### CODE REVIEW

File: include/arch/arm/aarch64/sys_io.h
Comment (galak) [inline include/arch/arm/aarch64/sys_io.h:26]: Might be good to add a comment in the source similar to the one in the commit message.

File: arch/arm/core/aarch64/Kconfig
Comment (carlocaione) [inline arch/arm/core/aarch64/Kconfig:64]: move the default value before the help

File: arch/arm/core/aarch64/header.S
Comment (carlocaione) [inline arch/arm/core/aarch64/header.S:13]: align the comment

File: include/arch/arm/aarch64/sys_io.h
Comment (carlocaione) [inline include/arch/arm/aarch64/sys_io.h:30]: The indentation is off for the changes in this file

**Review by carlocaione — COMMENTED** [review]
Xen support is really cool, thank you for this PR. Can you create a proper documentation for the xenvm board like https://github.com/zephyrproject-rtos/zephyr/tree/master/boards/arm/qemu_cortex_a53/doc just to have the commit message of f9ae96c0a2c0 dumped there?

**Comment by klogg** [comment]
@lorc please fix copyrights

File: boards/arm/xenvm/xenvm.dts
Comment (jgrall) [inline boards/arm/xenvm/xenvm.dts:9]: Does it mean Zephyr is not able to detect the amount of memory at runtime?

File: boards/arm/xenvm/xenvm.dts
Comment (jgrall) [inline boards/arm/xenvm/xenvm.dts:54]: What if a user want to run Zephyr with GICv3? Would we need to create a new Device-Tree?

File: boards/arm/xenvm/xenvm.dts
Comment (jgrall) [inline boards/arm/xenvm/xenvm.dts:79]: An SBSA driver would be able to drive a PL011 UART. But a PL011 driver may not be able to drive an SBSA UART because not all SBSA UART are PL011. For instance, the Xen emulation is a cut-down version.

So I am not sure this is correct to add the property ``arm,pl011``. Note that AFAICT, the toolstack doesn't expose ``arm,pl011``.

File: soc/arm/xenvm/Kconfig.soc
Comment (jgrall) [inline soc/arm/xenvm/Kconfig.soc:8]: Shouldn't this be selected by the user itself?

**Review by jgrall — COMMENTED** [review]
```

Remember to configure SYS_CLOCK_HW_CYCLES_PER_SEC for your machine. Xen
provides the correct value is DTB, but as zephyr can't parse device tree
in runtime, it can't get the correct value.
```

The correct way to find the clock frequency is by reading the system register ``CNTFRQ_EL0``. Only platform with broken firmware (e.g. they don't initialize the system register) would have the property in the device-tree.

So lets promote good behavior and not bad behavior in the commit message :).

File: drivers/serial/uart_pl011.c
Comment (jgrall) [inline drivers/serial/uart_pl011.c:425]: I think this change should be explained in the commit message.

File: drivers/serial/Kconfig.pl011
Comment (jgrall) [inline drivers/serial/Kconfig.pl011:30]: I don't really understand why a PL011 emulated by an hypervisor implies there are no need for configuration.

For instance, there are no promise that UART registers (such as IMCS, ICR) will be initializing a value expected by Zephyr. Imagine there was a firmware running before hand which configured the UART differently.

File: boards/arm/xenvm/xenvm.dts
Comment (lorc) [reply boards/arm/xenvm/xenvm.dts:9]: Yes. AFAIK, Zephyr uses static configuration.

File: boards/arm/xenvm/xenvm.dts
Comment (lorc) [reply boards/arm/xenvm/xenvm.dts:54]: Yes. Zephyr does not parses device tree at runtime. During compilation, special scripts generates header file from a provided DTS, with all needed definitions.

File: boards/arm/xenvm/xenvm.dts
Comment (lorc) [reply boards/arm/xenvm/xenvm.dts:79]: Yes, you are right.
I think right solution is to add new compatible value to existing pl011 driver in Zephyr.

File: drivers/serial/Kconfig.pl011
Comment (lorc) [reply drivers/serial/Kconfig.pl011:30]: Well, current vpl011 implementation in Xen lacks emulation of CR and LR_H registers.  Actually, I have a patch for Xen that handles reads and writes to those registers. But ignores what was written, so this is not a correct emulation.

On other hand, you are rigth and I should not omit configuration of IMCS and ICR from Zephyr side.

**Comment by lorc** [comment]
> The correct way to find the clock frequency is by reading the system register `CNTFRQ_EL0`. Only platform with broken firmware (e.g. they don't initialize the system register) would have the property in the device-tree.
>
> So lets promote good behavior and not bad behavior in the commit message :).

Sure :)

File: drivers/serial/Kconfig.pl011
Comment (bertrand-marquis) [reply drivers/serial/Kconfig.pl011:30]: We should try to fix this in xen instead, i agree with Julien here.
That would make it easier for other RTOS to run on top of Xen "unmodified"

Comment (lorc) [reply drivers/serial/Kconfig.pl011]: Bertrand, I agree with you that it would be great to have full PL011 emulation in Xen. But correct emulation requires a lot of efforts and I have no resources to implement it right now.

Instead, I added "SBSA mode"  to existing PL011 driver in Zephyr. I tried to align it with existing code, so I hope, this will be enough.

File: boards/arm/xenvm/xenvm.dts
Comment (lorc) [reply boards/arm/xenvm/xenvm.dts:56]: I added board manual. It describes how to update Zephyr configuration in case of any changes from Xen side.

**Comment by lorc** [comment]
I have addressed all the comments and pushed the new version

File: soc/arm/xenvm/mmu_regions.c
Comment (carlocaione) [inline soc/arm/xenvm/mmu_regions.c:15]: Probably this should be made more explicit but using `MT_xx` is frowned upon. Please switch to use one of the `MT_P_xx_U_xx` macros (`MT_P_RW_U_RW` probably in your case).

**Comment by carlocaione** [comment]
> I have addressed all the comments and pushed the new version

I still do not see the documentation though (see https://github.com/zephyrproject-rtos/zephyr/pull/30164#pullrequestreview-535617421)

**Comment by lorc** [comment]
> I still do not see the documentation though (see #30164 (review))

Sorry, forgot to add new files to the commit. Should be fine now.

File: boards/arm/xenvm/xenvm.dts
Comment (lorc) [reply boards/arm/xenvm/xenvm.dts:83]: This bothers me too, but Zephyr uses different DTS bindings and they require 4 cells in "interrupts" property:
https://github.com/zephyrproject-rtos/zephyr/blob/master/dts/bindings/interrupt-controller/arm%2Cgic.yaml

File: boards/arm/xenvm/xenvm.dts
Comment (lorc) [reply boards/arm/xenvm/xenvm.dts]: Yes, I know. But as I said, Zephyr uses different DTS bindings. This poses problems, when porting DTS from Linux, but...

File: drivers/serial/uart_pl011.c
Comment (lorc) [reply drivers/serial/uart_pl011.c]: Yes, this is working code. Probably, you missed my reply to Julien. Zephyr uses DTS for a built-time configuration. Special python script parses DTS file during build and generates a header file with all device tree data, which then is being accessed by `DT_` macros.

**Comment by lorc** [comment]
I can see that "buildkite/zephyr" hook failed. But it failed because it used all free disk space. I believe, this is not caused by my PR. Is there any way to re-run this hook?

**Comment by MrVan** [comment]
This is to make zephyr run as dom0 or domu? Not relevant to xen, I just enabled zephyr Jailhouse hypervisor inmate cell, and will enable zephyr as jailhouse root cell.

File: arch/arm/core/aarch64/header.S
Comment (MrVan) [inline arch/arm/core/aarch64/header.S:14]: extra `_`?

File: boards/arm/xenvm/doc/index.rst
Comment (MrVan) [inline boards/arm/xenvm/doc/index.rst:21]: why not GICv3?

File: boards/arm/xenvm/doc/index.rst
Comment (MrVan) [inline boards/arm/xenvm/doc/index.rst:49]: 8.32MHz, this is per SoC, not fixed.

File: boards/arm/xenvm/doc/index.rst
Comment (MrVan) [inline boards/arm/xenvm/doc/index.rst:64]: Since zephyr is used for RT, PV console is needed?

File: boards/arm/xenvm/doc/index.rst
Comment (MrVan) [inline boards/arm/xenvm/doc/index.rst:85]: I might not follow up, is `memore` right?

File: soc/arm/xenvm/plat_core.S
Comment (MrVan) [inline soc/arm/xenvm/plat_core.S:16]: This file is placeholder?

File: soc/arm/xenvm/mmu_regions.c
Comment (MrVan) [inline soc/arm/xenvm/mmu_regions.c:26]: MT_NS needed?

**Comment by beriberikix** [comment]
@sstabellini FYI. Might your RunX work be helpful/relevant here?

**Comment by sstabellini** [comment]
@beriberikix absolutely! We have already baremetal/unikernel apps running as containers with RunX, and RTOSes are next. Zephyr is a great candidate.

File: boards/arm/xenvm/doc/index.rst
Comment (lorc) [reply boards/arm/xenvm/doc/index.rst:21]: Because GICv2 is supported on all platforms.

File: boards/arm/xenvm/doc/index.rst
Comment (lorc) [reply boards/arm/xenvm/doc/index.rst:49]: Right. Maybe this part of documentation is not clear? I tried to emphasis that this setting should be changed for user's actual hardware.

File: boards/arm/xenvm/doc/index.rst
Comment (lorc) [reply boards/arm/xenvm/doc/index.rst:64]: I don't see any difference between PV console and SBSA console in this regard. Anyways, right now Xen does not provides any latency guarantees. 

This is why I created this PR in the first place: I'm trying to bring Xen to RT and I needed some domain that can be used for latency measurements.

File: boards/arm/xenvm/doc/index.rst
Comment (lorc) [reply boards/arm/xenvm/doc/index.rst:85]: Yes, thanks for the pointing out

File: arch/arm/core/aarch64/header.S
Comment (lorc) [reply arch/arm/core/aarch64/header.S:14]: Good catch! Thank you

File: soc/arm/xenvm/plat_core.S
Comment (lorc) [reply soc/arm/xenvm/plat_core.S:16]: Yes, actually it can be removed from the PR

File: soc/arm/xenvm/mmu_regions.c
Comment (lorc) [reply soc/arm/xenvm/mmu_regions.c:26]: Yes, I'll add it in the new version

**Comment by lorc** [comment]
Addressed comments, rebased and pushed a new version.

looks like @MrVan  also encountered issue with vector table alignment :)

File: boards/arm/xenvm/doc/index.rst
Comment (jgrall) [reply boards/arm/xenvm/doc/index.rst:21]: Not really. The GICv3 specification doesn't mandate support for GICv2.

Comment (lorc) [reply boards/arm/xenvm/doc/index.rst:21]: Oh, didn't knew that. 

Anyways,  I tested this PR with on RCAR Gen3 hardware, which have GICv2 and system timer frequency of 8.320 MHz. Hence the default values.

File: arch/arm/core/aarch64/header.S
Comment (carlocaione) [inline arch/arm/core/aarch64/header.S:16]: Spurious `P` in `HEADERP_PGSIZE`

File: arch/arm/core/aarch64/header.S
Comment (carlocaione) [inline arch/arm/core/aarch64/header.S:18]: ditto

File: soc/arm/xenvm/Kconfig.soc
Comment (carlocaione) [inline soc/arm/xenvm/Kconfig.soc:6]: This is not needed

File: arch/arm/core/aarch64/Kconfig
Comment (tejlmand) [inline arch/arm/core/aarch64/Kconfig:59]: this appears very specific for aarch64, should it be named: `AARCH64_IMAGE_HEADER` or `ARM64_IMAGE_HEADER` instead ?

File: drivers/interrupt_controller/Kconfig.gic
Comment (tejlmand) [inline drivers/interrupt_controller/Kconfig.gic:12]: based on the commit message, I still don't understand why prompt is needed here ?

This should not be configurable by the user, but something the board decides, from how I understand the commit message.

When you create: `zenvm` board, you may create the defconfig with:
`boards/arm/zenvm/Kconfig.defconfig
```
 config GIC_V1
	default y
```
and it will be enabled even for just the `zenvm` board.
This allows the board to select this setting, but not the user.

File: drivers/interrupt_controller/Kconfig.gic
Comment (tejlmand) [inline drivers/interrupt_controller/Kconfig.gic:19]: same as other promt comment.

File: drivers/interrupt_controller/Kconfig.gic
Comment (tejlmand) [inline drivers/interrupt_controller/Kconfig.gic:26]: same as other prompt comment.

File: arch/arm/core/aarch64/Kconfig
Comment (tejlmand) [inline arch/arm/core/aarch64/Kconfig:16]: same as other prompt comment.

File: arch/arm/core/aarch64/Kconfig
Comment (tejlmand) [inline arch/arm/core/aarch64/Kconfig:23]: same as other prompt comment.

File: boards/arm/xenvm/Kconfig.defconfig
Comment (tejlmand) [inline boards/arm/xenvm/Kconfig.defconfig:11]: You should probably add:
```suggestion

config GIC_V2
	default y

config CPU_CORTEX_A72=y
	default y
```

if you need, then you could make a board variant with the A53 SoC.

File: boards/arm/xenvm/xenvm_defconfig
Comment (tejlmand) [inline boards/arm/xenvm/xenvm_defconfig:4]: Those should be decided by the board, and not something the user can configure.

**Review by tejlmand — CHANGES_REQUESTED** [review]
The should not be a need for specifying prompts on `GIC_Vn` or `CPU_CORTEX_An`.

File: boards/arm/xenvm/Kconfig.defconfig
Comment (jgrall) [reply boards/arm/xenvm/Kconfig.defconfig:11]: I am a bit concerned this will clutter the number of xenvm variant in Zephyr. We would be looking to at least one board variant per CPU type. You can then multiply it by the number of options that a user can select in the VM configuration file (e.g gic version, device passthrough, RAM...).

Exposing these options to the user seems a better approach to me. Another approach that could be considered is providing a script that would be able to generate the board configuration based on the VM configuration.

Comment (tejlmand) [reply boards/arm/xenvm/Kconfig.defconfig:11]: except that such change will suddenly allow users to reconfigure the SoC on a real board.
That is something that should not be possible, as it may lead to all sorts of confusions later.

Also running Kconfig and change the SoC will most likely impact the generated device tree, and thus the build folder should probably be pristine cleaned.

With this PR, it also becomes possible to select two SoCs (or none), so it basically breaks all rules.

I see the need for more flexibility when working with VMs, and it is possible in the board Kconfig to add a prompt and in that way allow user to select a SoC.
However, such approach must be careful evaluated and tested, especially with regards to DTS.
First step would be to only allow users to select the SoC on first CMake invocation, similar to how it is only possible to select the board, but not change that selection without a pristine build.

Example in `boards/arm/zenvm/Kconfig.defconfig`
```
config CPU_CORTEX_A53
       bool "Allow reconfig (A53)"

config CPU_CORTEX_A72
       bool "Allow reconfig (A72)"
       depends on !CPU_CORTEX_A53
       default y
```

Comment (lorc) [reply boards/arm/xenvm/Kconfig.defconfig:11]: Well, I don't even sure that in case of VMs there is a need to divide configuration between SoC and board. Every VM is different in almost all aspects: CPU core type, number of cores, memory size, type of interrupt controller, timer frequency, set of peripherals. I can see only two constant things: SBSA UART and need for Linux image header, so Xen can load and run Zephyr.

Actually, even that is not constant because SBSA might be replaced with Xen virtual conso and Xen image loader supports ELF images (however, only at x86 right now).

So, maybe Zephyr does not need "xenvm" soc and "xenvm" board at all. Or need them as mere examples. You have seen that people already asked "why GICv2?", "why 8.320MHz?". Only answer I have is "because this is what needed on my platform".  

I have two ideas there: I can drop all patches related to "xenvm" and leave only patches with SBSA and Linux Image header, so any user that wants to run Zephyr on Xen can create own board with own configuration.

Or I can rename "xenvm" to "xenvm_rcar_gen3_example", because this is exactly what it is: example showing how to run Zephyr as Xen virtual guest on Renesas RCar Gen3 hardware. This will lift most of the questions about proposed configuration. 

Also, I love Julien's idea about scripts that creates board configuration based on input parameters. But I don't think that we have resources for this.

Comment (tejlmand) [reply boards/arm/xenvm/Kconfig.defconfig:11]: > Every VM is different in almost all aspects: CPU core type, number of cores, memory size, type of interrupt controller, timer frequency, set of peripherals. 

Yes, that is one of the advantages of using VMs for testing.
Unfortunately current board design was not having this is mind when started.

>  I can drop all patches related to "xenvm" 

I think it's great to have a starting point, even if it is fixed.
It makes it easier for users to use the board as base line and adjust it for their own need.

This is not so different from vendor A providing Board B, and some creates their own board with same SoC and using Board B as baseline for a new board, Board B_C.

So let's keep your `xenvm` proposal, but without changing the current way configs are handled.

User can then copy `xenvm` into their own repo and adjust the settings the need.
They can easily add that repo as additional board root.
https://docs.zephyrproject.org/latest/guides/modules.html#build-settings

> Also, I love Julien's idea about scripts that creates board configuration based on input parameters. But I don't think that we have resources for this.

Yes, let's look into this for the future.
But waiting for this should not prevent us from taking this PR in.

File: soc/arm/xenvm/Kconfig.soc
Comment (lorc) [reply soc/arm/xenvm/Kconfig.soc:6]: I tried to remove `select ARM` but then I'm getting the following error:

> warning: ARM64 (defined at arch/arm/Kconfig:12) has direct dependencies ARM with value n, but is currently being y-selected by the following symbols:
>  - BOARD_XENVM (defined at boards/arm/xenvm/Kconfig.board:4), with value y, direct dependencies SOC_XENVM && <choice> (value: y), and select condition SOC_XENVM && <choice> (value: y)
>  - SOC_XENVM (defined at soc/arm/xenvm/Kconfig.soc:4), with value y, direct dependencies <choice> (value: y), and select condition <choice> (value: y)

File: boards/arm/xenvm/xenvm_defconfig
Comment (carlocaione) [inline boards/arm/xenvm/xenvm_defconfig:11]: It's better to not hardcode these here. See #30443

File: boards/arm/xenvm/doc/index.rst
Comment (carlocaione) [inline boards/arm/xenvm/doc/index.rst:123]: I don't see any `zephyr.bin` as output of `west ..`, only `zephyr.elf`. Any missing steps here?

**Comment by carlocaione** [comment]
LGTM now, but this should be ACK-ed by XEN people as well.

**Comment by lorc** [comment]
@sstabellini, @jgrall, @bertrand-marquis, do you have any comments?

**Review by bertrand-marquis — COMMENTED** [review]
I did not have the chance to actually test that, but the changes are looking right to me.

**Comment by sstabellini** [comment]
Everything looks good to me.

Acked-by: Stefano Stabellini <sstabellini@kernel.org>

**Review by carlocaione — APPROVED** [review]

**Comment by lorc** [comment]
I'm not familiar with Zephyr contribution process... Are there any action needed from my side?

**Comment by carlocaione** [comment]
@lorc you need to rebase your PR because it has conflicts.

After that @tejlmand should review this again since he requested a change. And in general you have to wait for two approvals before the PR can be merged.

**Review by tejlmand — COMMENTED** [review]
Looking good. Thanks for the update.

Only a couple of small nits.

**Comment by lorc** [comment]
Sorry for the long pause.

I resolved @tejlmand's comments and rebased this PR onto current master. 
Looks like one patch was already taken into the mainline.

**Review by carlocaione — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Comment by nashif** [comment]
@ioannisg  please review

File: arch/arm/core/aarch64/header.S
Comment (carlocaione) [inline arch/arm/core/aarch64/header.S:11]: `PAGE_SIZE` was removed as part of #29551. `CONFIG_MMU_PAGE_SIZE` must be used here.

**Comment by lorc** [comment]
Rebased. Apart from `CONFIG_MMU_PAGE_SIZE` change, I replaced deprecated macro in PL011 driver and removed PSCI node from dts.

**Comment by carlocaione** [comment]
> and removed PSCI node from dts.

Oh, any reason why given that we have now a PSCI driver? #30911

**Comment by lorc** [comment]
> Oh, any reason why given that we have now a PSCI driver? #30911

I had some weird issues when I tried to boot Zephyr with PSCI node present. 

I invested some time in debugging and found the issue. Problem was in my setup: I use NFS to provide zephyr image for my target. Due to some caching quirks, client's view of image was different from server's and this caused random faults during the boot. 

So, I returned PSCI node back.

**Comment by nashif** [comment]
rebase please?

**Comment by lorc** [comment]
Rebased

**Review by nashif — APPROVED** [review]

**Pull request merged**
